mdoc/src/main/scala/mdoc/internal/worksheets/WorksheetProvider.scala

```diff
@@ -35,26 +35,55 @@ class WorksheetProvider(settings: Settings) {
     val compiler = ctx.compiler(instrumented)
     val rendered =
       MarkdownBuilder.buildDocument(compiler, reporter, sectionInputs, instrumented, input.path)
-    val decorations = for {
+    val diagnostics = reporter.diagnostics.map(d => d: i.Diagnostic).toSeq
+
+    val evaluatedStatements = (for {
       section <- rendered.sections.iterator
       statement <- section.section.statements
-    } yield renderDecoration(statement)
+    } yield renderDecoration(statement)).toList
+
+    val exportableEvaluation =
+      if (diagnostics.exists(_.severity() == i.DiagnosticSeverity.Error)) {
+        ju.Optional.empty[String]
+      } else {
+        ju.Optional.of(zipStatements(input, evaluatedStatements))
+      }
 
     EvaluatedWorksheet(
-      reporter.diagnostics.map(d => d: i.Diagnostic).toSeq.asJava,
-      decorations
+      diagnostics.asJava,
+      evaluatedStatements
         .filterNot(_.summary.isEmpty)
         .map(d => d: i.EvaluatedWorksheetStatement)
-        .toList
         .asJava,
       instrumented.fileImports.map(_.toInterface).asJava,
       instrumented.scalacOptionImports.map(_.value).asJava,
       compiler.classpathEntries.asJava,
       instrumented.dependencies.toSeq.asJava,
-      instrumented.repositories.toSeq.asJava
+      instrumented.repositories.toSeq.asJava,
+      exportableEvaluation
     )
   }
 
+  private def zipStatements(
```

Comment:
Is there a reason this logic needs to be implemented on the mdoc side? It would be nice if clients can implement their own rendering.

Reply:
Ah, I think I may have misunderstood you when we spoke briefly on discord about this:

> :thinking: for sharing an evaluated worksheet, do you think you'd ever do that outside of Metals? My initial thought was for that just to be in Metals, not in mdoc

> I meant that if sbt-mdoc has a way to automatically generate the evaluates output for all worksheet files then it would be cool to expose that feature inside Metals as an action to export a single worksheet into the clipboard

https://discord.com/channels/632642981228314653/634377575455653898/766211091574947841

So I understood that as you saying it should be done in mdoc 😬 

So in general, I agree, with the current way this is implemented for Metals use, it's actually not necessary to do anything in mdoc to get the copied worksheet. We can cache the `EvaluatedWorksheet` when we get it, and then on demand, create the exportable worksheet and give it to the user. So you're totally right.

The only reason I was also focused on doing it in here was so that sbt-mdoc or a even the mdoc cli if need be could just use this as well to evaluate a worksheet and then grab the pretty output.

Maybe a better solution is to implement this part fully in Metals? And then tackle the other issue complete separately? wdyt?

Reply:
I think it would be useful to have it in mdoc and be able to test out worksheets separately. We were talking about soemthing like this with Vinz from Scala Center. He was trying to figure out if he can automatically test their educational material.

Reply:
I think making an additional call to mdoc to get this though probably doesn't make a ton of sense, especially seeing that we have everything we need after the evaluated worksheet comes back to create the exportable pretty output. I originally had it in as another field primarily because I didn't want to make another call to create the output. I'm sort of leaning towards closing this, implementing it fully on the Metals side, and then tackling the issue of testing worksheets separately.

Reply:
I agree it's best to avoid duplicate evaluation to export the evaluation into a string.

The reason I lean towards rendering this on the client side is because there may be different needs for how to render the output. I am concerned that we'll end up needing to support configuration options to tweak how mdoc exports the evaluation.

There are still good reasons to keep the rendering logic in the mdoc side

* it's easier to test inside the mdoc repo
* the rendering logic can be non-trivial to implement and most clients may be happy with the defaults

I'm open for either approach, I'll leave the decision to y'all

Reply:
So I'm going to go ahead and close this, and then implement straight in Metals. Thanks a lot for the feedback guys. 🤜🏼 🤛🏼